api/app/weather_models/rdps_filename_marshaller.py

```diff
@@ -0,0 +1,87 @@
+"""
+Reading and writing of RDPS URLs, based on the specification here:
+https://eccc-msc.github.io/open-data/msc-data/nwp_rdps/readme_rdps-datamart_en/#filename-nomenclature
+"""
+
+from datetime import datetime
+
+# Canadian Meterological Centre constant
+CMC = "CMC"
+# RDPS constant
+REG = "reg"
+# Accumulating precipitation constant
+APCP = "APCP"
+# Surface level type
+SFC = "SFC"
+# Level 0
+LEVEL = "0"
+# Polar-stereographic projection at 10km resolution
+PS10KM = "ps10km"
+# Grib2 file type
+GRIB2 = "grib2"
+# Filename delimiter
+DELIMITER = "_"
+# Possible run hours
+RUN_HOURS = ["00", "06", "12", "18"]
+# Possible forecast hours
+FORECAST_HOURS = [f"{hour:03d}" for hour in list(range(0, 84))]
+
+
+def parse_rdps_filename(url: str):
+    """Parse and return the forecast start date and run hour from the RDPS grib url."""
+    tokens = url.split(DELIMITER)
+    assert tokens[0] == CMC
+    assert tokens[1] == REG
+    assert tokens[2] == APCP
+    assert tokens[3] == SFC
+    assert tokens[4] == LEVEL
+    assert tokens[5] == PS10KM
+
+    # utc datetime in YYYYMMDDHH format
+    assert len(tokens[6]) == 10
+    forecast_start_date = tokens[6][:-2]
+    run_hour = tokens[6][-2:]
+    assert run_hour in RUN_HOURS
+
+    # forecast hour and grib2 file ext
+    assert len(tokens[7]) == 10
+    final_tokens = tokens[7].split(".")
+    assert len(final_tokens) == 2
+    # strip off constant 'P', get forecast hours triplet
+    forecast_hour = final_tokens[0][-3:]
+    assert forecast_hour in FORECAST_HOURS
+
+    return (forecast_start_date, run_hour, forecast_hour)
+
+
+def check_compose_invariants(forecast_start_date: datetime, run_hour: int, forecast_hour: int):
+    """Explode if any of these assertions fail"""
+    assert forecast_start_date.tzinfo is not None
+    assert int(forecast_start_date.utcoffset().total_seconds()) == 0
+    assert f"{forecast_hour:03d}" in FORECAST_HOURS
+
+
+def compose_rdps_filename(forecast_start_date: datetime, run_hour: int, forecast_hour: int):
+    """Compose and return an RDPS url given a forecast start date, run hour and forecast hour.
+    An RDPS url has a run hour in [00, 12].
+    """
+    check_compose_invariants(forecast_start_date, run_hour, forecast_hour)
+    # a model run must have a correct run hour
+    assert f"{run_hour:02d}" in RUN_HOURS
+
+    return (
```

Comment:
nit: I think this return statement is the same as the one in `compose_computed_rdps_filename`. You could make the assertion here and return a call to `compose_computed_rdps_filename` to avoid the code duplication.

Reply:
Thanks, fixed in https://github.com/bcgov/wps/pull/3749/commits/b009e4ded00e4655a8907935785afe8ca5da4a00